src/utils/project-manipulation.js

```diff
@@ -32,12 +38,155 @@ export function filterProjectFields(project, edit = false) {
+  }
+
+  activity.timeframe = {
+    start: toISOFormat(activity.timeframe.start),
+    end: toISOFormat(activity.timeframe.end),
+  };
+
+  if (activity.application_deadline != null) {
+    activity.application_deadline = toISOFormat(activity.application_deadline);
+  }
+
   delete activity.project;
   delete activity.applications;
+  delete activity.existing_application;
   delete activity.internal;
+  delete activity.vacant_spots;
+}
+
+/* Prepare the activity object to be processed on the frontend
+   by converting dates to Date objects. */
+export function prepareAfterBackend(activity) {
+  activity.timeframe = {
+    start: new Date(activity.timeframe.start),
+    end: new Date(activity.timeframe.end),
+  };
+
+  if (activity.application_deadline != null) {
+    activity.application_deadline = new Date(activity.application_deadline);
+  }
+}
+
+/* Determine the index to insert the activity into the list
+   knowing its position among non-internal activities. */
+export function determineInsertionIndex(activities, position) {
+  let insertionIndex = 0;
+  while (position && insertionIndex < activities.length) {
+    if (!activities[insertionIndex].internal) {
+      position--;
+    }
+    insertionIndex++;
+  }
+  return insertionIndex;
+}
+
+/* Synchronize the internal representation of the activities (with types)
+   with the activity list from the backend. */
+export function synchronizeActivityLists(internalList, backendActivities) {
+  let internalListIdx = 0;
+  for (let activity of backendActivities.filter(activity => !activity.internal)) {
+    let activityProcessed = false;
+    while (!activityProcessed) {
+      // If the internal list is over, copy the backend activity to the end
+      // Happens on the initial population of the activityList
+      if (internalListIdx >= internalList.length) {
+        let copy = copyActivity(activity);
+        copy._type = ActivityTypes.DISPLAY;
+        prepareAfterBackend(copy);
+        internalList.splice(internalListIdx, 0, copy);
+        internalListIdx++;
+        activityProcessed = true;
+        break;
+      }
+
+      // If the internalList cursor is on a deletion marker, remove it from the list
+      // Happens when an activity deletion was confirmed by a dialog
+      if (internalList[internalListIdx]._type === ActivityTypes.DELETION_MARKER) {
+        internalList.splice(internalListIdx, 1);
+        continue;
+      }
+
+      // If the internalList cursor is on a new activity form, let it be
+      // Happens when there is an open form that hasn't been saved yet
+      if (internalList[internalListIdx]._type === ActivityTypes.NEW) {
+        internalListIdx++;
+        continue;
+      }
+
+      // If the internalList cursor is on a displayed activity,
+      // don't change anything – it is already that same activity
+      if (internalList[internalListIdx]._type === ActivityTypes.DISPLAY) {
+        internalListIdx++;
+        activityProcessed = true;
+        break;
```

Comment:
Either the variable or the `break;` statements are useless

Reply:
Yes, the variable was introduced for the readability of this while loop. Removing it would make the loop `while (true)`, and it's not clear what's the terminating condition for that loop. Just done to preserve sanity of future coders

Reply:
So that makes the `break;` unnecessary

Reply:
No, the `break` ensures that the loop execution doesn't go further.

Reply:
and why do you need that? It seems that all if-conditions are mutually exclusive anyways. They are all checking if one property is `===` something

Reply:
The `internalListIdx` gets modified in process, which means these conditions wouldn't be checking the same thing. So they are not mutually exclusive in that sense